pkg/testcase/upgradenodereplacement.go

```diff
@@ -0,0 +1,570 @@
+	if ip == "" {
+		return shared.ReturnLogError("ip not sent\n")
+	}
+
+	if delNodeErr := shared.DeleteClusterNode(ip); delNodeErr != nil {
+		shared.LogLevel("error", "error deleting server: %w\n", delNodeErr)
+
+		return delNodeErr
+	}
+	shared.LogLevel("info", "Node IP deleted from the cluster: %s\n", ip)
+
+	err := a.DeleteInstance(ip)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func replaceAgents(
+	cluster *factory.Cluster,
+	a *aws.Client,
+	serverLeaderIp, token, version string,
+	newExternalAgentIps, newPrivateAgentIps []string,
+) error {
+	if token == "" {
+		return shared.ReturnLogError("token not sent\n")
+	}
+
+	if len(newExternalAgentIps) == 0 || len(newPrivateAgentIps) == 0 {
+		return shared.ReturnLogError("externalIps or privateIps empty\n")
+	}
+
+	if err := deleteAgents(a, cluster); err != nil {
+		shared.LogLevel("error", "error deleting agent: %w\n", err)
+
+		return err
+	}
+
+	for i, externalIp := range newExternalAgentIps {
+		privateIp := newPrivateAgentIps[i]
+
+		joinErr := joinAgent(cluster.Config.Product, serverLeaderIp, token, version, externalIp, privateIp)
+		if joinErr != nil {
+			shared.LogLevel("error", "error joining agent: %w\n", joinErr)
+
+			return joinErr
+		}
+	}
+
+	return nil
+}
+
+func deleteAgents(a *aws.Client, c *factory.Cluster) error {
+	for _, i := range c.AgentIPs {
+		if deleteNodeErr := shared.DeleteClusterNode(i); deleteNodeErr != nil {
+			shared.LogLevel("error", "error deleting agent: %w\n", deleteNodeErr)
+
+			return deleteNodeErr
+		}
+		shared.LogLevel("info", "Node IP deleted from the cluster: %s\n", i)
+
+		err := a.DeleteInstance(i)
+		if err != nil {
+			return err
+		}
+		shared.LogLevel("info", "Instance IP deleted from cloud provider: %s\n", i)
+	}
+
+	return nil
+}
+
+func joinAgent(product, serverIp, token, version, selfExternalIp, selfPrivateIp string) error {
+	cmd, parseErr := buildJoinCmd(product, agent, serverIp, token, version, selfExternalIp, selfPrivateIp)
+	if parseErr != nil {
+		return shared.ReturnLogError("error parsing join commands: %w\n", parseErr)
+	}
+
+	if joinErr := joinNode(cmd, selfExternalIp); joinErr != nil {
+		return shared.ReturnLogError("error joining node: %w\n", joinErr)
+	}
+
+	return nil
+}
+
+func joinNode(cmd, ip string) error {
+	if cmd == "" {
+		return shared.ReturnLogError("cmd not sent\n")
+	}
+	if ip == "" {
+		return shared.ReturnLogError("server IP not sent\n")
+	}
+
+	res, err := shared.RunCommandOnNode(cmd, ip)
+	if err != nil {
+		return shared.ReturnLogError("error joining node: %w\n", err)
+	}
+	res = strings.TrimSpace(res)
+	if strings.Contains(res, "service failed") {
+		shared.LogLevel("error", "join node response: %s\n", res)
+
+		return shared.ReturnLogError("error joining node: %s\n", res)
+	}
+
+	delay := time.After(40 * time.Second)
+	// delay not meant to wait if node is joined, but rather to give time for all join process to complete under the hood
+	<-delay
```

Comment:
Same as other comment -- do we need this? If we really do, less complicated to just use `time.Sleep`

Reply:
already asnwered